### ISSUE TITLE
[Infrastructure Request] DataDog: HAProxy configuration for Console Services team

### ISSUE THREAD

**Issue by LindseySaari** [department-of-veterans-affairs/va.gov-team#42065]
### Team name

Console Services

### Application name

DataDog SLO report

### GitHub repository

https://app.datadoghq.com/dashboard/utj-23b-f7v/vsp---external-service-performance-indicators?from_ts=1654017039030&to_ts=1654020639030&live=true

https://grafana.vfs.va.gov/d/pEgVdRlZk/external-service-performance-indicators?orgId=1
### Product owner

Thomas Delrue

### Product manager

Nathan Peterson

### DevOps resource

None :cries:

### Related issue

https://github.com/orgs/department-of-veterans-affairs/projects/547/views/1?filterQuery=-label%3AEpic+slo

### Infrastructure needed

[More details in the Epic here](https://github.com/orgs/department-of-veterans-affairs/projects/547/views/1?filterQuery=dog)

The Platform Console Services Team is in the process of moving the SLO report from Grafana to DataDog. Part of the SLO report is to display percentile-based latency for each external service separated into buckets (50th, 90th, and 99th percentiles). The HaProxy metrics come from a [DataDog integration](https://app.datadoghq.com/account/settings#integrations/haproxy). We've exhausted our attempts to put the correct config in place for the haproxy. ([#11291](https://github.com/department-of-veterans-affairs/devops/pull/11291), [#11314](https://github.com/department-of-veterans-affairs/devops/pull/11314) and [#11330](https://github.com/department-of-veterans-affairs/devops/pull/11330)) We will need some support from the Infrastructure team to put the correct config in place to obtain the percentile based latency metrics for haproxy.

Questions:
1. How do we currently get Prometheus metrics for Haproxy?
2. Could we send the raw Data for Haproxy instead of using the Integration?
3. How are we currently seeing some percentile-based metrics for vets-api? We don't even have the dogstatsd client implemented. [See search here.](https://app.datadoghq.com/metric/summary?filter=external_http)

Other:
- I [sent a flare](https://docs.datadoghq.com/agent/troubleshooting/send_a_flare/?tab=agentv6v7#pagetitle) to DataDog support containing our current haproxy config. [Chat here.]( https://help.datadoghq.com/hc/en-us/requests/817023)
- [Chat between Infra and Console Services Engineers](https://dsva.slack.com/archives/CJYRZK2HH/p1653055329376499)

TLDR; We need some assistance to get the correct metrics sent from the haproxy into DataDog for percentile based latency.

### Acceptance Criteria

- [ ] _TBD_

**Comment by jhouse-solvd** [comment]
@LindseySaari - Thanks for submitting this. We are working on the DataDog migration this sprint, so it might make sense to wait on this until the next sprint. However, I'll discuss it with the team and get back to you soon.

**Comment by jhouse-solvd** [comment]
Planning to discuss w/ the Platform Infrastructure team on Friday.

**Comment by jhouse-solvd** [comment]
This issue is scheduled for the current sprint. ☺️ 

We'll reach out to clarify requirements and update the status of the ticket as we go. 

Let us know if you have any questions.

**Comment by mleclerc00** [comment]
Currently the haproxy integration is setup correctly and shooting metrics off to datadog.  Where the problem comes in the differences in functionality that grafana and datadog provide.

As we can see by looking at the following dashboard in grafana the query is being run on the `haproxy_server_total_time_average_seconds` metric.  https://grafana.vfs.va.gov/d/pEgVdRlZk/external-service-performance-indicators?editPanel=1061&orgId=1&var-data_source=Prometheus%20(Production)&var-backend=All&var-code=All&var-component=All&from=now-7d&to=now

`quantile(0.99, haproxy_server_total_time_average_seconds{proxy=~"$backend"})`

This particular metric as provided by the haproxy prometheus endpoint is a gauge 

![image](https://user-images.githubusercontent.com/76792875/177589934-afcc5408-5ec5-4328-ae97-b2880eb4988f.png)

In grafana, we can use the listed aggregation operators here to run calculations on any metric type. One of these being the quantile aggregator. https://prometheus.io/docs/prometheus/latest/querying/operators/#aggregation-operators

Where this breaks down is the fact that datadog does not provide the same aggregation operators or any functions on gauge type metrics that will provide the same results (as far as I can tell after a lot of testing).  

https://docs.datadoghq.com/dashboards/querying/#aggregate-and-rollup
https://docs.datadoghq.com/dashboards/functions/

In datadog land we cannot calculate a percentile of any metric type, it needs to be a histogram or distribution.  

https://docs.datadoghq.com/metrics/types/?tab=histogram&tabs=histogram#metric-types
https://docs.datadoghq.com/metrics/types/?tab=histogram&tabs=distribution#metric-types

Coming back to haproxy we can also see that there have been issues related to this exact question in the past, but was ultimately dismissed 

https://github.com/haproxy/haproxy/issues/699

Coming across this also showed that it is an issue that others have come by before.  Side note, I'm not necessarily suggesting this since it could be a rabbit hole in its own.  But there are articles that reference people using fluentd to scrape logs and create their own histogram metrics.  An example being https://medium.com/@tom.fawcett/extracting-useful-duration-metrics-from-haproxy-prometheus-fluentd-2be9832ff702

Ultimately, based on the findings above I don't think there is a quick 1:1 migration of the quantile aggregation from grafana to datadog with the metrics being provided by haproxy

**Comment by jhouse-solvd** [comment]
@LindseySaari & @npeterson54 - The team did a pretty deep dive into this (see Matt's comment above). We didn't find a way to "migrate" the current percentile-based functionality to Datadog. 

There might be a different way for your team to implement this by adding instrumentation to the Vets-API application, or perhaps by consulting with Datadog engineers directly. 

Let us know how we can support you. We're happy to continue to conversation.

**Comment by mleclerc00** [comment]
closing this for the time being since there isn't anything further we can do for percentile metrics via the integration itself
